### ISSUE TITLE
`HttpCookieStore` incorrectly rejects cookies for domains that are an IPv6 address

### ISSUE THREAD

**Issue by lorban** [eclipse/jetty.project#9665]
**Jetty version(s)**
12

**Description**
```java
HttpCookieStore.add(
    URI.create("http://[::1]"),
    HttpCookie.build("n", "v").domain("[::1]").build()
);
```

returns `false` while it does return `true` if an IPv4 is used:

```java
HttpCookieStore.add(
    URI.create("http://127.0.0.1"),
    HttpCookie.build("n", "v").domain("127.0.0.1").build()
);
```

This breaks cookies handling when Jetty is accessed via an URL using an IPv6 as the hostname.

**Comment by joakime** [comment]
If we had a `URIUtil.isHostName(String)` we could have the logic be different for Hostname vs IP-Literal.

Example: ad5f8c4d151214c656a325c3144e700abbe9551a
This is just a proof of concept, not efficient.

``` java
    /**
     * Test if the token is a valid hostname and not an IP-Literal.
     *
     * @param token the token to check
     * @return true if token is a Host name, not an IP-Literal
     */
    public static boolean isHostName(String token)
    {
        Pattern patIPv4 = Pattern.compile("\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}");
        if (patIPv4.matcher(token).matches())
            return false;
        else
            return isValidHostRegisteredName(token);
    }
```

And test case

``` java
    @ParameterizedTest
    @CsvSource(textBlock = """
        127.0.0.1       , false
        ::1             , false
        [::1]           , false
        localhost       , true
        localhost.local , true
        sub.example.com , true
        """)
    public void testIsHostName(String token, boolean expected)
    {
        assertEquals(expected, URIUtil.isHostName(token));
    }
```

**Comment by joakime** [comment]
Interestingly, the "Domain Matching" (5.1.3) at https://datatracker.ietf.org/doc/html/rfc6265#section-5.1.2  seems to indicate that having a cookie from an IP-Literal is not valid.

But tests (jetty demo webapp) on Chrome show that Cookies with IP-Literals (be it IPv4 or IPv6) are both seem to be supported by Chrome.

**Comment by joakime** [comment]
Some more testing on Chrome.

From a location of `http://127.0.0.1:8888/cookie/`, the following `Set-Cookie` is reported as valid to Chrome.

```
Set-Cookie: aaa=bbb; Domain=127.0.0.1; Expires=Wed, 19-Apr-2023 19:10:01 GMT; Max-Age=60
```

But from location `http://[::1]:8888/cookie/` the following `Set-Cookie` is reported as invalid and ignored.

```
Set-Cookie: aaa=bbb; Domain=::1; Expires=Wed, 19-Apr-2023 20:01:33 GMT; Max-Age=60
```

This is likely due to the missing `[...]` brackets around the IPv6-Literal.

If I add them back the cookie becomes valid again.

```
Set-Cookie: aaa=bbb; Domain=[::1]; Expires=Wed, 19-Apr-2023 20:18:03 GMT; Max-Age=60
```

**Comment by joakime** [comment]
On review of HttpCookieStore there was a TODO ...

https://github.com/eclipse/jetty.project/blob/43e8c83e242269f9de4d6f91ae104257cac89892/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCookieStore.java#L145-L151

This is kinda complicated.

The official list of domain suffixes is maintained by mozilla at https://publicsuffix.org/
It is updated quite frequently at their github page https://github.com/publicsuffix/list ([`/public_suffix_list.dat`](https://github.com/publicsuffix/list/blob/master/public_suffix_list.dat))

Not sure how we would go about maintaining this active changing list.

**Comment by sbordet** [comment]
Fixed by #10465.

**Comment by sbordet** [comment]
Fixed by #10465.
